### ISSUE TITLE
Inline assembly

### ISSUE THREAD

**Issue by Amanieu** [rust-lang/compiler-team#280]
# TL;DR

- Implement the new `asm!` syntax proposed in https://github.com/rust-lang/rfcs/pull/2873.
- The new syntax is much more user-friendly and has significantly more checks in rustc, which should make it impossible to trigger an ICE, unlike the old implementation.

# Links and Details

In https://github.com/rust-lang/rust/pull/71007 we deprecated the old `asm!` macro and told users to use `llvm_asm!` instead. The next step is to replace `asm!` with a macro implementing the new syntax proposed in the RFC.

The implementation PR has design notes on how the inline assembly passes through the various stages of the compiler (AST, HIR, MIR, etc).

Implementation PR: https://github.com/rust-lang/rust/pull/69171
New RFC thread: https://github.com/rust-lang/rfcs/pull/2873
Old RFC thread (now closed): https://github.com/rust-lang/rfcs/pull/2850

# Mentors or Reviewers

@nagisa has already performed an initial review of the implementation PR.

**Comment by nikomatsakis** [comment]
@rustbot second 

I second this proposal -- basically, everything here looks "roughly like" I would expect it to look, and @nagisa has already done an initial review, so we have a dedicated reviewer.

**Comment by nikomatsakis** [comment]
FCP expired, closing the issue!
